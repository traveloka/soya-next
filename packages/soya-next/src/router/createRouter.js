import Router from "express/lib/router";
import cookieMiddleware from "universal-cookie-express";
import createLocaleMiddleware from "./createLocaleMiddleware";
import ensureRedirect from "../utils/ensureRedirect";
import parseRedirectionPath from "../utils/parseRedirectionPath";
import { toPath } from "../utils/locale";

const defaultOptions = {
  routes: {},
  redirects: {}
};

export default (
  app,
  {
    basePath,
    routes = defaultOptions.routes,
    redirects = defaultOptions.redirects,
    defaultLocale,
    siteLocales,
    compression,
    whoami = {}
  } = defaultOptions
) => {
  const router = Router();
  const handle = app.getRequestHandler();
  if (!app.dev) {
    router.use(require("compression")(compression));
  }

  const newRedirects = Object.keys(redirects).reduce((newRedirects, from) => {
    const redirect = redirects[from];
    const newRoute = routes[redirect.to];
    newRedirects[from] = {
      page: (newRoute && newRoute.page) || redirect.to,
      ...redirect
    };
    return newRedirects;
  }, {});
  router.use((req, res, next) => {
    req.redirects = newRedirects;
    next();
  });
  router.use(cookieMiddleware());
  if (defaultLocale && siteLocales) {
    router.use(createLocaleMiddleware({ defaultLocale, siteLocales }));
  }
  if (basePath) {
    router.use((req, res, next) => {
      let test;
      let exclude;
      if (typeof basePath === "string") {
        test = basePath;
      } else {
        ({ test, exclude } = basePath);
      }
      const excludes = (app.dev
        ? ["/_next/webpack-hmr", "/_next/on-demand-entries-ping"]
        : []
      ).concat(exclude || []);
      if (
        (excludes.length && excludes.includes(req.url)) ||
        req.url.startsWith(test)
      ) {
        req.url = req.url.replace(new RegExp(`^${test}/*`), "/");
        return next();
      }
      app.render404(req, res);
    });
  }
  Object.keys(newRedirects).forEach(from => {
    const { method, status, to } = ensureRedirect(newRedirects[from]);
    router[method.toLowerCase()](from, (req, res) => {
      const localeSegment = toPath(req.locale, defaultLocale);
      const redirectionPath = parseRedirectionPath(
        localeSegment + to,
        req.params
      );
      res.redirect(status, redirectionPath);
    });
  });
  Object.keys(routes).forEach(path => {
    const { method = "GET", page } = routes[path];
    router[method.toLowerCase()](path, (req, res) => {
      app.render(req, res, page, Object.assign({}, req.query, req.params));
    });
  });
  router.get("/whoami", (req, res) => {
    const uptime = process.uptime();
    res.json({
      nodeId: process.env.NODE_GROUP,
      uptime: `${Math.floor(uptime / 86400)}d, ${Math.floor(
        (uptime % 86400) / 3600
      )}h, ${Math.floor((uptime % 3600) / 60)}m, ${Math.floor(uptime % 60)}s`,
      ...whoami
    });
  });
  router.get("*", (req, res) => handle(req, res));
  return router;
};
