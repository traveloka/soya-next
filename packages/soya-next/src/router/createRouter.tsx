import { Router } from "express";
import cookieMiddleware from "universal-cookie-express";
import createLocaleMiddleware from "./createLocaleMiddleware";
import ensureRedirect from "../utils/ensureRedirect";
import parseRedirectionPath from "../utils/parseRedirectionPath";
import { toPath } from "../utils/locale";

import type { Request, Response } from "express";
import type { ServerResponse } from "http";
import type Server from "next/dist/next-server/server/next-server";
import type {
  SoyaNextConfig,
  SoyaNextRedirectConfig,
  SoyaNextServerConfig,
} from "../types";

declare module "next/dist/next-server/server/next-server" {
  export default interface Server {
    /**
     * @deprecated `dev` property does not exist anymore
     * since next 7.
     */
    dev?: boolean;
  }
}

declare module "express-serve-static-core" {
  export interface Router extends Record<string, any> {}
}

export interface CreateRouterOption {
  basePath?: SoyaNextConfig["basePath"];
  compression?: SoyaNextServerConfig["compression"];
  defaultLocale?: SoyaNextConfig["defaultLocale"];
  redirects?: SoyaNextConfig["redirects"];
  routes?: SoyaNextConfig["routes"];
  siteLocales?: SoyaNextConfig["siteLocales"];
  whoami?: SoyaNextConfig["whoami"];
}

export default function createRouter(
  app: Server,
  options?: CreateRouterOption
) {
  const {
    basePath,
    routes = {},
    redirects = {},
    defaultLocale,
    siteLocales,
    compression,
    whoami = {},
  } = options || {};
  const router = Router();
  const handle = app.getRequestHandler();
  if (!app.dev) {
    router.use(require("compression")(compression));
  }

  const newRedirects = Object.keys(redirects).reduce<
    Record<string, SoyaNextRedirectConfig>
  >((newRedirects, from) => {
    const redirect = redirects[from];
    const newRoute = routes[redirect.to];
    newRedirects[from] = {
      page: (newRoute && newRoute.page) || redirect.to,
      ...redirect,
    };
    return newRedirects;
  }, {});
  router.use((req, res, next) => {
    req.redirects = newRedirects;
    next();
  });
  router.use(cookieMiddleware());
  if (defaultLocale && siteLocales) {
    router.use(createLocaleMiddleware({ defaultLocale, siteLocales }));
  }
  if (basePath) {
    router.use((req, res, next) => {
      let test: string | undefined;
      let exclude: string | string[] | undefined;
      if (typeof basePath === "string") {
        test = basePath;
      } else {
        ({ test, exclude } = basePath);
      }
      const excludes = (app.dev
        ? ["/_next/webpack-hmr", "/_next/on-demand-entries-ping"]
        : []
      ).concat(exclude || []);
      if (
        (excludes.length && excludes.includes(req.url)) ||
        req.url.startsWith(String(test))
      ) {
        req.url = req.url.replace(new RegExp(`^${test}/*`), "/");
        return next();
      }
      // INFO: Don't know why express `Response` type error here while
      // it already extends the `http.ServerResponse` class.
      app.render404(req, res as ServerResponse);
    });
  }
  Object.keys(newRedirects).forEach(from => {
    const { method, status, to } = ensureRedirect(newRedirects[from]);
    router[method.toLowerCase()](from, (req: any, res: any) => {
      const localeSegment = toPath(req.locale, defaultLocale);
      const redirectionPath = parseRedirectionPath(
        localeSegment + to,
        req.params
      );
      res.redirect(status, redirectionPath);
    });
  });
  Object.keys(routes).forEach(path => {
    const { method = "GET", page } = routes[path];
    router[method.toLowerCase()](path, (req: Request, res: Response) => {
      app.render(
        req,
        res as ServerResponse,
        page,
        Object.assign({}, req.query, req.params)
      );
    });
  });
  router.get("/whoami", (req, res) => {
    const uptime = process.uptime();
    res.json({
      nodeId: process.env.NODE_GROUP,
      uptime: `${Math.floor(uptime / 86400)}d, ${Math.floor(
        (uptime % 86400) / 3600
      )}h, ${Math.floor((uptime % 3600) / 60)}m, ${Math.floor(uptime % 60)}s`,
      ...whoami,
    });
  });
  router.get("*", (req, res) => handle(req, res as ServerResponse));
  return router;
}
