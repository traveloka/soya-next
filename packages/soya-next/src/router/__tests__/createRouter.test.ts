import createRouter from "../createRouter";

jest.mock("express/lib/router", () =>
  jest.fn(() => {
    const middlewares: any[] = [];
    const routes: any[] = [];
    return {
      middlewares,
      routes,
      get: jest.fn((path, handler) => {
        routes.push({ path, handler });
      }),
      use: jest.fn(middleware => {
        middlewares.push(middleware);
      }),
    };
  })
);

describe("createRouter", () => {
  function createFakeNext(dev: boolean = false) {
    const handle = jest.fn();
    return {
      dev,
      dir: ".",
      dist: ".next",
      getRequestHandler: jest.fn(() => handle),
      handle,
      render: jest.fn(),
      render404: jest.fn(),
      serveStatic: jest.fn(),
    };
  }

  describe("production", () => {
    it("should create gzip enabled router", () => {
      const app = createFakeNext();
      const router = createRouter(app as any);
      // @ts-ignore
      expect(router.use.mock.calls.length).toBe(3);
    });

    it("should create router with base path and exclude a path", () => {
      const app = createFakeNext();
      const router = createRouter(app as any, {
        basePath: {
          test: "/base",
          exclude: "/healthcheck",
        },
      });
      const req = { url: "/healthcheck" };
      const res = { json: jest.fn() };
      const next = jest.fn();

      router.middlewares[3](req, res, next);
      expect(req.url).toBe("/healthcheck");
      expect(next).toBeCalled();
    });
  });

  describe("non-production", () => {
    it("should create basic router", async () => {
      const app = createFakeNext(true);
      const router = createRouter(app as any);
      const req = {
        headers: {},
        params: {
          path: "/react.png",
        },
      };
      const res = { json: jest.fn() };
      const next = jest.fn();

      // redirects middleware
      router.middlewares[0](req, res, next);
      expect(req).toHaveProperty("redirects");
      expect(next).toBeCalled();

      // universal cookies middleware
      router.middlewares[1](req, res, next);
      expect(req).toHaveProperty("universalCookies");
      expect(next.mock.calls.length).toBe(2);

      // whoami
      router.routes[0].handler(req, res);
      expect(res.json).toBeCalled();

      // next handler
      await router.routes[1].handler(req, res);
      // @ts-ignore
      expect(app.handle).toBeCalled();

      // @ts-ignore
      expect(router.use.mock.calls.length).toBe(2);
      // @ts-ignore
      expect(router.get.mock.calls.length).toBe(2);
      expect(app.getRequestHandler).toBeCalled();
    });

    it("should create router with custom routes", async () => {
      const app = createFakeNext(true);
      const routes = {
        "/p/:id": {
          page: "/post",
        },
      };
      const router = createRouter(app as any, { routes });
      await router.routes[0].handler(
        {
          query: { locale: "en-id" },
          params: { id: 1 },
        },
        {}
      );
      expect(app.render).toBeCalled();
      expect(app.render.mock.calls[0]).toMatchSnapshot();
      // @ts-ignore
      expect(router.get.mock.calls.length).toBe(3);
    });

    it("should create router with redirection", () => {
      const app = createFakeNext(true);
      const defaultLocale = "id-id";
      const routes = {
        "/p/:id": {
          page: "/post",
        },
      };
      const redirects = {
        "/tentang": {
          to: "/about",
        },
        "/post/:id": {
          to: "/p/:id",
        },
      };
      const router = createRouter(app as any, {
        defaultLocale,
        routes,
        redirects,
      });
      const res = {
        redirect: jest.fn(),
      };
      router.routes[0].handler({}, res);
      expect(res.redirect).toBeCalled();
      expect(res.redirect.mock.calls[0]).toMatchSnapshot();
      router.routes[1].handler(
        {
          locale: {
            language: "en",
            country: "id",
          },
          params: { id: 1 },
        },
        res
      );
      expect(res.redirect.mock.calls[1]).toMatchSnapshot();
      // @ts-ignore
      expect(router.get.mock.calls.length).toBe(5);
    });

    it("should create locale aware router", () => {
      const app = createFakeNext(true);
      const router = createRouter(app as any, {
        defaultLocale: "id-id",
        siteLocales: ["id-id", "en-id"],
      });
      // @ts-ignore
      expect(router.use.mock.calls.length).toBe(3);
    });

    it("should create router with base path", () => {
      const app = createFakeNext(true);
      const router = createRouter(app as any, {
        basePath: "/base",
      });
      const req = { url: "/base" };
      const res = { json: jest.fn() };
      const next = jest.fn();

      router.middlewares[2](req, res, next);
      expect(req.url).toBe("/");
      expect(next).toBeCalled();

      router.middlewares[2]({ url: "/" }, res, next);
      expect(app.render404).toBeCalled();
    });

    it("should create router with base path and exclude some paths", () => {
      const app = createFakeNext(true);
      const router = createRouter(app as any, {
        basePath: {
          test: "/base",
          exclude: ["/healthcheck", "/whoami"],
        },
      });
      const req = { url: "/whoami" };
      const res = { json: jest.fn() };
      const next = jest.fn();

      router.middlewares[2](req, res, next);
      expect(req.url).toBe("/whoami");
      expect(next).toBeCalled();
    });
  });
});
