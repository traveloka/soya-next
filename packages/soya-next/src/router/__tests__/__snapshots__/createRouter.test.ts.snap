// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createRouter non-production should create router with custom routes 1`] = `
Array [
  Object {
    "params": Object {
      "id": 1,
    },
    "query": Object {
      "locale": "en-id",
    },
  },
  Object {},
  "/post",
  Object {
    "id": 1,
    "locale": "en-id",
  },
]
`;

exports[`createRouter non-production should create router with redirection 1`] = `
Array [
  301,
  "/about",
]
`;

exports[`createRouter non-production should create router with redirection 2`] = `
Array [
  301,
  "/en/p/1",
]
`;
